src/_guides/language/effective-dart/design.md

```diff
@@ -1005,79 +1091,260 @@ var names = people.map((Person person) {
+
+Dart has two notations for defining a named typedef for a function type. The
+original syntax looks like:
+
+{:.bad-style}
+<?code-excerpt "misc/lib/effective_dart/design_bad.dart (old-typedef)"?>
+{% prettify dart %}
+typedef int Comparison<T>(T a, T b);
+{% endprettify %}
+
+It has a couple of problems:
+
+*   There is no way to assign a name to a *generic* function type. In the above
+    example, the typedef itself is generic. If you reference `Comparison` in
+    your code, without a type argument, you implicitly get the function type
+    `int Function(dynamic, dynamic)`, *not* `int Function<T>(T, T)`. This
+    doesn't come up in practice often, but it matters in certain corner cases.
+
+*   A single identifier in a parameter is interpreted as the parameter's *name*,
+    not its *type*. Given:
+
+    {:.bad-style}
+    <?code-excerpt "misc/lib/effective_dart/design_bad.dart (typedef-param)"?>
+    {% prettify dart %}
+    typedef bool TestNumber(num);
+    {% endprettify %}
+
+    Most users expect this to be a function type that takes a `num` and returns
+    `bool`. It is actually a function type that takes *any* object (`dynamic`)
+    and returns `bool`. The parameter's *name* (which isn't used for anything
+    except documentation in the typedef) is "num". This has been a
+    long-standing source of errors in Dart.
+
+The new syntax looks like:
+
+{:.good-style}
+<?code-excerpt "misc/lib/effective_dart/design_good.dart (new-typedef)"?>
+{% prettify dart %}
+typedef Comparison = int Function<T>(T, T);
+{% endprettify %}
+
+The new syntax can express anything the old syntax could express and more, and
+lacks the error-prone misfeature where a single identifier is treated as the
+parameter's name. The same function type syntax after the `=` in the typedef is
+also allowed anywhere a type annotation may appear, giving us a single
+consistent way to write function types anywhere in a program.
+
+The old typedef syntax is still supported to avoid breaking existing code, but
+is deprecated.
+
+
+### PREFER inline function types over typedefs.
+
+In Dart 1, if you wanted to use a function type for a field, variable, or
+generic type argument, you had to first define a typedef for it. Dart 2 supports
+a function type syntax that can be used anywhere a type annotation is allowed:
+
+{:.good-style}
+<?code-excerpt "misc/lib/effective_dart/design_good.dart (function-type)"?>
+{% prettify dart %}
+class FilteredObservable {
+  // In a field:
+  final bool Function(Event) _predicate;
+
+  // In a generic:
+  final List<void Function(Event)> _observers;
+
+  FilteredObservable(this._predicate, this._observers);
+
+  // In a return type:
+  void Function(Event) notify(Event event) {
+    if (!_predicate(event)) return null;
+
+    // In a local variable:
+    void Function(Event) last;
+    for (var observer in _observers) {
+      observer(event);
+      last = observer;
+    }
+
+    return last;
+  }
+}
+{% endprettify %}
```

Comment:
Personally, I find the `In a ...` comments distracting.

Reply:
Do you? I thought it would be good to spell out all of the cases and what each case represents. (In particular, I think using a function type as a type argument or a return type is non-obvious.)

Reply:
I prefer the highlights to the comments. It pulls in my eye and tickles my curiosity: "why are there so many highlighted regions? Oh, they are all different usages, cool!" :)

@kwalrath what's your take on this: leave the comments out or put them back in? (The comments would describe what kind of function type annotation it is: field, generic type argument, etc.)

Reply:
I definitely like the highlights, but it's easy to miss what the importance of each example is. The comments seem like a nice way to highlight that.

Reply:
Yeah, I like the highlights too. Killed the comments.